### ISSUE TITLE
Allow to specify the result type of a sum

### ISSUE THREAD

**Issue by lindahua** [JuliaLang/julia#5311]
Currently, the `sum` function use the element type (except for the boolean case) of the input array as result type. This is sometimes not a desirable choice:

``` julia
function mysum(R, x)
    s = zero(R)
    for v in x; s += v; end
    return s
end

julia> x = uint8(rand(1:10, 50));

julia> sum(x)
0x0000000000000122

# to prevent overflow
julia> mysum(Int, x)
290

julia> a = rand(Float32, 10^6);

julia> sum(float64(a)) - sum(a)
-0.018809685949236155

# to get higher accuracy
julia> sum(float64(a)) - mysum(0.0, a)
1.1641532182693481e-10
```

My proposal is to add a method like `sum(R, x)` where `R` is the user-specifed type. This won't break any existing codes, but provide the user additional choices.

**Comment by oscardssmith** [comment]
Given the existence of 18423, should this be closed?

**Comment by kmsquire** [comment]
@oscardssmith typically issues are closed only after relevant pull requests are merged.

**Comment by yuyichao** [comment]
Or when the two issues are duplicate of each other (or when one covers the other one). These two issues seems to be closely related but are not quite the same. (A fix should probably take both into account though.)

**Comment by martinholters** [comment]
The original issue seems to be solved, `sum(OutputType, x)` does work:
```Julia
julia> a = rand(Float32, 10^6);

julia> sum(Float64.(a)) - sum(a)
-0.020957350730895996

julia> sum(Float64.(a)) - sum(Float64,a)
0.0
```

Close or change title?

PS: I haven't verified there is a test for this; we might want to do so before closing.

**Comment by KristofferC** [comment]
Works now.
